### ISSUE TITLE
[saltnado rest] local requests hang indefinitely on glob / grain queries

### ISSUE THREAD

**Issue by baweaver** [saltstack/salt#21707]
I'm making a rubygem handler for Salt REST API, and for some reason whenever I use a glob or grain query it'll stall indefinitely.

I'm using Typhoeus to make HTTP requests.

The request code looks like this:
```ruby
client.lowstate.local(fun: 'test.ping', tgt: '*')

# Request
#<Typhoeus::Request:0x007ff92b10b4a0
 @base_url="localhost:8000/",
 @options=
  {:method=>:post,
   :body=>{:client=>"local", :tgt=>"*", :fun=>"test.ping", :expr_form=>"glob"},
   :headers=>
    {"User-Agent"=>"Typhoeus - https://github.com/typhoeus/typhoeus",
     :Accept=>"application/json",
     :ContentType=>"application/json",
     "X-Auth-Token"=>"token"},
   :maxredirs=>50},
 @original_options=
  {:method=>:post,
   :body=>{:client=>"local", :tgt=>"*", :fun=>"test.ping", :expr_form=>"glob"},
   :headers=>
    {:Accept=>"application/json",
     :ContentType=>"application/json",
     "X-Auth-Token"=>"token"}}>

# body: "client=local&tgt=*&fun=test.ping&expr_form=glob"
# url:  "localhost:8000/"
```

I can do the same thing on ``local_async`` and it'll work but I have no way of knowing when the job is actually done.

Any ideas?

**Comment by rallytime** [comment]
@baweaver I am not well-versed in the ways of saltnado (though @whiteinge or maybe even @jacksontj are good bets here), but would something like [salt.utils.http.query](https://github.com/saltstack/salt/blob/3c76046691c44d50330c6b70f8555211a792c2e9/salt/utils/http.py#L68) could be useful for you here?

**Comment by baweaver** [comment]
I'll look into it and let you know what I find.

**Comment by jacksontj** [comment]
From our chat on IRC it sounds like the issue might be the ruby client (since it appears to work in curl). I double checked and this case has tests around it (https://github.com/saltstack/salt/blob/develop/tests/integration/netapi/rest_tornado/test_app.py#L86), so its probably not saltnado. @baweaver will keep us updated as he digs in :)

**Comment by baweaver** [comment]
After talking with @jacksontj, we've agreed to try upgrading the server.

Pertinent machine information:
```
Salt Version: salt 2014.7.2 (Helium)
ZeroMQ Version: 3.2.4
```

Trying this on Curl: (blanked actual urls and tokens)
```
curl -svi http://localhost:8000  -H "Accept: application/x-yaml"  -H "X-Auth-Token: my_token" -d client=local -d tgt='*.url.net' -d fun='test.ping'

 * Adding handle: conn: 0x7fa1fa003a00
* Adding handle: send: 0
* Adding handle: recv: 0
* Curl_addHandleToPipeline: length: 1
* - Conn 0 (0x7fa1fa003a00) send_pipe: 1, recv_pipe: 0
* About to connect() to localhost port 8000 (#0)
*   Trying ::1...
* Connected to localhost (::1) port 8000 (#0)
> POST / HTTP/1.1
> User-Agent: curl/7.30.0
> Host: localhost:8000
> Accept: application/x-yaml
> X-Auth-Token: my_token
> Content-Length: 52
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 52 out of 52 bytes

# stalls indefinitely until 5 minute timeout here
```

On the server I see events trying to return:
```
2015-03-17 02:18:50,870 [salt.master                                ][INFO    ] Got return from url.net for job 20150317021849103091
2015-03-17 02:18:50,871 [salt.utils.event                           ][DEBUG   ] Sending event - data = {'fun_args': [], 'jid': '20150317021849103091', 'return': True, 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2015-03-17T
02:18:50.870535', 'fun': 'test.ping', 'id': 'url.net'}
2015-03-17 02:18:50,871 [salt.utils.event                           ][DEBUG   ] Sending event - data = {'fun_args': [], 'jid': '20150317021849103091', 'return': True, 'retcode': 0, 'success': True, 'cmd': '_return', '_stamp': '2015-03-17T
02:18:50.870792', 'fun': 'test.ping', 'id': 'url.net'}
2015-03-17 02:18:50,948 [salt.master                                ][INFO    ] AES payload received with command _return
```

This will only happen on local with a glob or grain expression. A specific hostname will give no errors.

Thanks again for your time @jacksontj!

**Comment by baweaver** [comment]
It looks like 2014.7.2 (Helium) is the latest.

**Comment by baweaver** [comment]
In the mean time, and as a hackaround solution, I've been using the async jobs and retrieving the results within a few seconds. Is there any way to see if a job is done?

**Comment by jacksontj** [comment]
I am unable to reproduce this locally still. I was looking into this, and I'm a bit interested with the 5m timeout there. From the code I'm looking at the salt-api enforces no such timeout-- which lead me to the problem. The issue is that the publisher says the glob matched "N" minions, but not all N are available. This means the salt-api will actually wait indefinitely for the return to show up on the master event bus-- which isn't great ;) This fix (#21755) adds the same timeout that the CLI has-- which should mean you get the same returns (more or less) that you would have gotten if you ran it on the CLI-- which is I think the goal here.

**Comment by rallytime** [comment]
Thanks for the discussion and for the fix! I believe the pull request listed above is for the 2015.2 branch, but the fix that is related to the 2014.7 branch is #21755 (which for some reason isn't showing up in the usual, nicely formatted, github way) - just wanted to link it here for ease of use.

@baweaver Can you give that pull request a try and let us know if you run into any problems with it?

**Comment by baweaver** [comment]
I'll look into it and ping the owner to see what we can get done. Thanks again for your time! I'll post back with what we find.
